File: Parallel & Distributed Computing/Project/MESI/mesi.py
# Constants
NUM_CACHES = 4
ARRAY_SIZE = 20

# Define cache line states
INVALID = 'I'
SHARED = 'S'
MODIFIED = 'M'
EXCLUSIVE = 'E'

# Shared array (main memory)
main_memory = [0] * ARRAY_SIZE

# Cache arrays
cache = [[None] * ARRAY_SIZE for _ in range(NUM_CACHES)]
cache_state = [[INVALID] * ARRAY_SIZE for _ in range(NUM_CACHES)]


def display_system_state():
    print("Main Memory:")
    print(main_memory)
    print("\nCaches:")
    for i in range(NUM_CACHES):
        print(f"Cache {i}:")
        print(cache[i])
        print(f"Cache State:")
        print(cache_state[i])
    print("\n")


def read_from_cache(cache_id, index):
    if cache_state[cache_id][index] == INVALID:
        cache[cache_id][index] = main_memory[index]
        cache_state[cache_id][index] = EXCLUSIVE
        for j in range(NUM_CACHES):
            if j != cache_id and cache_state[j][index] == EXCLUSIVE:
                cache_state[j][index] = SHARED
                cache_state[cache_id][index] = SHARED
    elif cache_state[cache_id][index] == SHARED:
        pass
    elif cache_state[cache_id][index] == EXCLUSIVE:
        pass
    elif cache_state[cache_id][index] == MODIFIED:
        pass
    print(f"Cache {cache_id} reads from index {index}")


def write_to_cache(cache_id, index, value):
    if cache_state[cache_id][index] == INVALID or cache_state[cache_id][index] == SHARED:
        cache[cache_id][index] = value
        main_memory[index] = value
        cache_state[cache_id][index] = MODIFIED
        for j in range(NUM_CACHES):
            if j != cache_id:
                cache_state[j][index] = INVALID
    elif cache_state[cache_id][index] == EXCLUSIVE:
        cache[cache_id][index] = value
        main_memory[index] = value
        cache_state[cache_id][index] = MODIFIED
    elif cache_state[cache_id][index] == MODIFIED:
        cache[cache_id][index] = value
        main_memory[index] = value
    print(f"Cache {cache_id} writes to index {index}")


def main():
    while True:
        display_system_state()
        choice = input("Select a cache (0-3) or 'q' to quit: ")
        if choice.lower() == 'q':
            break
        if not choice.isdigit() or int(choice) not in range(NUM_CACHES):
            print("Invalid choice. Please select a cache (0-3).")
            continue
        cache_id = int(choice)
        print(f"Cache {cache_id} State:")
        print(cache_state[cache_id])
        index = int(input("Select an index to read/write (0-99): "))
        if index not in range(ARRAY_SIZE):
            print("Invalid index. Please select an index (0-99).")
            continue
        action = input("Select action: 'r' for read, 'w' for write: ").lower()
        if action == 'r':
            read_from_cache(cache_id, index)
        elif action == 'w':
            value = int(input("Enter value to write: "))
            write_to_cache(cache_id, index, value)
        else:
            print("Invalid action. Please select 'r' for read or 'w' for write.")


if __name__ == "__main__":
    main()

File: Parallel & Distributed Computing/Project/MESI/Mesi.cpp
#include <iostream>
using namespace std;

// Enum to represent cache states
enum CacheState
{
    INVALID,
    SHARED,
    EXCLUSIVE,
    MODIFIED
};

// Global variables representing cache state and data
int cacheData = 0;
CacheState cacheState = INVALID;

void readFromMemory()
{
    // Simulate reading from main memory
    cacheData = 100;        // Example initial data
    cacheState = EXCLUSIVE; // Initially in EXCLUSIVE state
    cout << "Read from memory (EXCLUSIVE)" << endl;
}

void readFromCache()
{
    // Simulate reading from cache
    if (cacheState == INVALID)
    {
        readFromMemory();
    }
    else
    {
        cout << "Read from cache ";
        switch (cacheState)
        {
        case SHARED:
            cout << "(SHARED)" << endl;
            break;
        case EXCLUSIVE:
            cout << "(EXCLUSIVE)" << endl;
            break;
        case MODIFIED:
            cout << "(MODIFIED)" << endl;
            break;
        }
    }
}

void writeToMemory(int newData)
{
    // Simulate writing to main memory
    cacheData = newData;
    cacheState = EXCLUSIVE; // Move to EXCLUSIVE state
    cout << "Write to memory (EXCLUSIVE)" << endl;
}

void writeToCache(int newData)
{
    // Simulate writing to cache
    if (cacheState == INVALID)
    {
        writeToMemory(newData);
    }
    else
    {
        cacheData = newData;
        switch (cacheState)
        {
        case SHARED:
            cacheState = MODIFIED; // Move to MODIFIED state
            cout << "Write to cache (MODIFIED)" << endl;
            break;
        case EXCLUSIVE:
            cacheState = MODIFIED; // Move to MODIFIED state
            cout << "Write to cache (MODIFIED)" << endl;
            break;
        case MODIFIED:
            cout << "Write to cache (MODIFIED)" << endl;
            break;
        }
    }
}

void invalidateCache()
{
    cacheState = INVALID;
    cout << "Cache invalidated" << endl;
}

void printCacheState()
{
    cout << "Cache State: ";
    switch (cacheState)
    {
    case INVALID:
        cout << "INVALID" << endl;
        break;
    case SHARED:
        cout << "SHARED" << endl;
        break;
    case EXCLUSIVE:
        cout << "EXCLUSIVE" << endl;
        break;
    case MODIFIED:
        cout << "MODIFIED" << endl;
        break;
    }
}

int main()
{
    // Simulating cache operations
    readFromCache();   // Cache miss, fetch from memory (EXCLUSIVE)
    printCacheState(); // EXCLUSIVE

    writeToCache(200);
    printCacheState(); // MODIFIED

    readFromCache();   // Cache hit, read from cache (MODIFIED)
    printCacheState(); // MODIFIED

    invalidateCache();
    printCacheState(); // INVALID

    readFromCache();   // Cache miss, fetch from memory (EXCLUSIVE)
    printCacheState(); // EXCLUSIVE

    return 0;
}
